### PR TITLE
Tree Widget: Use type imports

### PR DESCRIPTION
**Pull request by saskliutas**
Closes https://github.com/iTwin/viewer-components-react/issues/470

Added eslint rule to force consistent usage of type imports.
Switched to using JSX without importing React.

### CODE REVIEW

File: packages/itwin/tree-widget/.eslintrc.js [hunk 1]
```diff
@@ -6,12 +6,15 @@ require("@rushstack/eslint-patch/modern-module-resolution");
 
 module.exports = {
   plugins: ["@itwin"],
-  extends: "plugin:@itwin/ui",
+  extends: ["plugin:@itwin/ui", "plugin:react/jsx-runtime"],
```
Comment (aruniverse) [inline packages/itwin/tree-widget/.eslintrc.js:9]: will this cause issues for apps? (i have no idea, but im in favor of removing the react import if possible)

File: packages/itwin/tree-widget/src/components/trees/models-tree/ModelsVisibilityHandler.ts [hunk 2]
```diff
@@ -3,18 +3,23 @@
...
 import { TreeWidget } from "../../../TreeWidget";
 import { toggleAllCategories } from "../CategoriesVisibilityUtils";
-import { IVisibilityHandler, VisibilityChangeListener, VisibilityStatus } from "../VisibilityTreeEventHandler";
 import { CachingElementIdsContainer } from "./Utils";
 
+import type { TreeNodeItem } from "@itwin/components-react";
+import type { Id64String } from "@itwin/core-bentley";
+import type { IModelConnection, Viewport } from "@itwin/core-frontend";
+import type { ECClassGroupingNodeKey, GroupingNodeKey, Keys} from "@itwin/presentation-common";
+import type { IFilteredPresentationTreeDataProvider} from "@itwin/presentation-components";
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/components/trees/models-tree/ModelsVisibilityHandler.ts:20]:
```suggestion
import type { ECClassGroupingNodeKey, GroupingNodeKey, Keys } from "@itwin/presentation-common";
import type { IFilteredPresentationTreeDataProvider } from "@itwin/presentation-components";
```

File: packages/itwin/tree-widget/src/test/IModelUtils.ts [hunk 3]
```diff
@@ -3,11 +3,11 @@
...
 
-import { Id64String } from "@itwin/core-bentley";
-import {
-  BisCodeSpec, CategoryProps, Code, GeometricElement3dProps, IModel, PhysicalElementProps, RelatedElementProps, SubCategoryProps,
-} from "@itwin/core-common";
-import { TestIModelBuilder } from "@itwin/presentation-testing";
+import { BisCodeSpec, Code, IModel } from "@itwin/core-common";
+
+import type { Id64String } from "@itwin/core-bentley";
+import type { CategoryProps, GeometricElement3dProps, PhysicalElementProps, RelatedElementProps, SubCategoryProps} from "@itwin/core-common";
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/test/IModelUtils.ts:9]:
```suggestion
import type { CategoryProps, GeometricElement3dProps, PhysicalElementProps, RelatedElementProps, SubCategoryProps } from "@itwin/core-common";

```

File: packages/itwin/tree-widget/src/test/TreeWidgetComponent.test.tsx [hunk 4]
```diff
@@ -3,17 +3,18 @@
...
 import { BeEvent } from "@itwin/core-bentley";
-import { IModelConnection } from "@itwin/core-frontend";
 import { cleanup, render, waitFor } from "@testing-library/react";
-import { TreeDefinition, TreeWidgetComponent } from "../tree-widget-react";
+import { TreeWidgetComponent } from "../tree-widget-react";
 import { TreeWidget } from "../TreeWidget";
 import { TestUtils } from "./TestUtils";
 
+import type { IModelConnection } from "@itwin/core-frontend";
+import type { TreeDefinition} from "../tree-widget-react";
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/test/TreeWidgetComponent.test.tsx:16]:
```suggestion
import type { TreeDefinition } from "../tree-widget-react";
```

File: packages/itwin/tree-widget/src/test/trees/CategoriesVisibilityUtils.test.ts [hunk 5]
```diff
@@ -3,15 +3,16 @@
...
-import {
-  IModelApp, IModelConnection, NoRenderApp, PerModelCategoryVisibility, ScreenViewport, SpatialViewState, ViewManager, Viewport, ViewState,
-} from "@itwin/core-frontend";
+import { SubCategoryAppearance } from "@itwin/core-common";
+import { IModelApp, NoRenderApp, PerModelCategoryVisibility } from "@itwin/core-frontend";
 import { enableCategory, enableSubCategory, toggleAllCategories } from "../../components/trees/CategoriesVisibilityUtils";
 import { TestUtils } from "../TestUtils";
 
+import type { ECSqlReader} from "@itwin/core-common";
+import type { IModelConnection, ScreenViewport, SpatialViewState, ViewManager, Viewport, ViewState} from "@itwin/core-frontend";
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/test/trees/CategoriesVisibilityUtils.test.ts:14]:
```suggestion
import type { ECSqlReader } from "@itwin/core-common";
import type { IModelConnection, ScreenViewport, SpatialViewState, ViewManager, Viewport, ViewState } from "@itwin/core-frontend";
```

File: packages/itwin/tree-widget/src/test/trees/Common.ts [hunk 6]
```diff
@@ -3,12 +3,16 @@
...
 import { CheckBoxState } from "@itwin/core-react";
-import { ECClassGroupingNodeKey, ECInstancesNodeKey, InstanceKey, StandardNodeTypes } from "@itwin/presentation-common";
-import { PresentationTreeNodeItem } from "@itwin/presentation-components";
+import { StandardNodeTypes } from "@itwin/presentation-common";
+
+import type { PropertyDescription, PropertyValue} from "@itwin/appui-abstract";
+import type { TreeModelNode } from "@itwin/components-react";
+import type { Id64String } from "@itwin/core-bentley";
+import type { ECClassGroupingNodeKey, ECInstancesNodeKey, InstanceKey} from "@itwin/presentation-common";
+import type { PresentationTreeNodeItem } from "@itwin/presentation-components";
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/test/trees/Common.ts:15]:
```suggestion
import type { PropertyDescription, PropertyValue } from "@itwin/appui-abstract";
import type { TreeModelNode } from "@itwin/components-react";
import type { Id64String } from "@itwin/core-bentley";
import type { ECClassGroupingNodeKey, ECInstancesNodeKey, InstanceKey } from "@itwin/presentation-common";
import type { PresentationTreeNodeItem } from "@itwin/presentation-components";
```

File: packages/itwin/tree-widget/src/test/trees/category-tree/CategoriesTree.test.tsx [hunk 7]
```diff
@@ -5,28 +5,33 @@
...
-import { VisibilityChangeListener } from "../../../components/trees/VisibilityTreeEventHandler";
 import {
   addDrawingCategory, addDrawingGraphic, addModel, addPartition, addPhysicalObject, addSpatialCategory, addSubCategory,
 } from "../../IModelUtils";
 import { mockPresentationManager, mockViewport, TestUtils } from "../../TestUtils";
 
+import type { TreeNodeItem } from "@itwin/components-react";
+import type { Id64String } from "@itwin/core-bentley";
+import type { IModelConnection, SpatialViewState, ViewManager, Viewport } from "@itwin/core-frontend";
+import type { ECInstancesNodeKey, Node, NodePathElement} from "@itwin/presentation-common";
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/test/trees/category-tree/CategoriesTree.test.tsx:29]:
```suggestion
import type { ECInstancesNodeKey, Node, NodePathElement } from "@itwin/presentation-common";
```

File: packages/itwin/tree-widget/src/test/trees/category-tree/CategoriesVisibilityHandler.test.ts [hunk 8]
```diff
@@ -4,23 +4,25 @@
...
+  CategoryVisibilityHandler, hideAllCategories, invertAllCategories, showAllCategories, useCategories,
 } from "../../../components/trees/category-tree/CategoryVisibilityHandler";
-import * as categoriesVisibilityUtils from "../../../components/trees/CategoriesVisibilityUtils";
 import { mockViewport } from "../../TestUtils";
 
+import type { Id64String} from "@itwin/core-bentley";
+import type { IModelConnection, ViewManager, Viewport, ViewState } from "@itwin/core-frontend";
+import type { ECInstancesNodeKey} from "@itwin/presentation-common";
+import type { PresentationTreeNodeItem } from "@itwin/presentation-components";
+import type { CategoryInfo, CategoryVisibilityHandlerParams} from "../../../components/trees/category-tree/CategoryVisibilityHandler";
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/test/trees/category-tree/CategoriesVisibilityHandler.test.ts:24]:
```suggestion
import type { Id64String } from "@itwin/core-bentley";
import type { IModelConnection, ViewManager, Viewport, ViewState } from "@itwin/core-frontend";
import type { ECInstancesNodeKey } from "@itwin/presentation-common";
import type { PresentationTreeNodeItem } from "@itwin/presentation-components";
import type { CategoryInfo, CategoryVisibilityHandlerParams } from "../../../components/trees/category-tree/CategoryVisibilityHandler";
```

File: packages/itwin/tree-widget/src/test/trees/imodel-content-tree/IModelContentTree.test.tsx [hunk 9]
```diff
@@ -26,6 +23,11 @@ import {
 } from "../../IModelUtils";
 import { mockPresentationManager, TestUtils } from "../../TestUtils";
 
+import type { TreeNodeItem } from "@itwin/components-react";
+import type { Id64String } from "@itwin/core-bentley";
+import type { IModelConnection} from "@itwin/core-frontend";
+import type { ECInstancesNodeKey} from "@itwin/presentation-common";
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/test/trees/imodel-content-tree/IModelContentTree.test.tsx:29]:
```suggestion
import type { IModelConnection } from "@itwin/core-frontend";
import type { ECInstancesNodeKey } from "@itwin/presentation-common";
```

File: packages/itwin/tree-widget/src/test/trees/models-tree/ModelsTree.test.tsx [hunk 10]
```diff
@@ -5,28 +5,33 @@
...
+import { ModelsTreeNodeType } from "../../../components/trees/models-tree/ModelsVisibilityHandler";
 import { createRuleset } from "../../../components/trees/models-tree/Utils";
-import { VisibilityChangeListener } from "../../../components/trees/VisibilityTreeEventHandler";
 import { addModel, addPartition, addPhysicalObject, addSpatialCategory, addSpatialLocationElement } from "../../IModelUtils";
 import { deepEquals, mockPresentationManager, mockViewport, TestUtils } from "../../TestUtils";
 import { createCategoryNode, createElementClassGroupingNode, createElementNode, createKey, createModelNode, createSubjectNode } from "../Common";
-import { IModel } from "@itwin/core-common";
+
+import type { TreeNodeItem } from "@itwin/components-react";
+import type { IModelConnection} from "@itwin/core-frontend";
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/test/trees/models-tree/ModelsTree.test.tsx:30]:
```suggestion
import type { IModelConnection } from "@itwin/core-frontend";
```

File: packages/itwin/tree-widget/src/test/trees/models-tree/ModelsTreeComponent.test.tsx [hunk 11]
```diff
@@ -3,16 +3,18 @@
...
-import { IModelApp, NoRenderApp, Viewport } from "@itwin/core-frontend";
-import sinon from "sinon";
-import { mockViewport, TestUtils } from "../../TestUtils";
-import { expect } from "chai";
 import * as modelsVisibilityHandler from "../../../components/trees/models-tree/ModelsVisibilityHandler";
-import { BeEvent } from "@itwin/core-bentley";
+import { ModelsTreeComponent } from "../../../tree-widget-react";
+import { mockViewport, TestUtils } from "../../TestUtils";
+
+import type { ModelInfo} from "../../../tree-widget-react";
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/test/trees/models-tree/ModelsTreeComponent.test.tsx:16]:
```suggestion
import type { ModelInfo } from "../../../tree-widget-react";
```

File: packages/itwin/tree-widget/src/test/trees/models-tree/ModelsVisibilityHandler.test.ts [hunk 12]
```diff
@@ -4,22 +4,30 @@
...
-import * as categoriesVisibilityUtils from "../../../components/trees/CategoriesVisibilityUtils";
+
+import type { Id64String} from "@itwin/core-bentley";
+import type { ECSqlReader} from "@itwin/core-common";
+import type { IModelConnection, Viewport, ViewState, ViewState3d } from "@itwin/core-frontend";
+import type { ECInstancesNodeKey } from "@itwin/presentation-common";
+import type { IFilteredPresentationTreeDataProvider, PresentationTreeNodeItem } from "@itwin/presentation-components";
+import type { IModelHierarchyChangeEventArgs, PresentationManager } from "@itwin/presentation-frontend";
+import type { ModelsVisibilityHandlerProps} from "../../../components/trees/models-tree/ModelsVisibilityHandler";
+import type { ModelInfo } from "../../../tree-widget-react";
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/test/trees/models-tree/ModelsVisibilityHandler.test.ts:30]:
```suggestion
import type { Id64String } from "@itwin/core-bentley";
import type { ECSqlReader } from "@itwin/core-common";
import type { IModelConnection, Viewport, ViewState, ViewState3d } from "@itwin/core-frontend";
import type { ECInstancesNodeKey } from "@itwin/presentation-common";
import type { IFilteredPresentationTreeDataProvider, PresentationTreeNodeItem } from "@itwin/presentation-components";
import type { IModelHierarchyChangeEventArgs, PresentationManager } from "@itwin/presentation-frontend";
import type { ModelsVisibilityHandlerProps } from "../../../components/trees/models-tree/ModelsVisibilityHandler";
import type { ModelInfo } from "../../../tree-widget-react";
```

File: packages/itwin/tree-widget/src/components/TreeWidgetComponent.tsx [hunk 13]
```diff
@@ -4,14 +4,17 @@
...
+import { useEffect, useState } from "react";
 import { useActiveIModelConnection } from "@itwin/appui-react";
-import { SelectableContent, SelectableContentDefinition, SelectableContentProps } from "@itwin/components-react";
-import { IModelConnection } from "@itwin/core-frontend";
+import { SelectableContent } from "@itwin/components-react";
 import { FillCentered } from "@itwin/core-react";
 import { ProgressLinear } from "@itwin/itwinui-react";
 import { TreeWidget } from "../TreeWidget";
 
+import type { PropsWithChildren} from "react";
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/components/TreeWidgetComponent.tsx:14]:
```suggestion
import type { PropsWithChildren } from "react";
```

File: packages/itwin/tree-widget/src/components/TreeWidgetUiItemsProvider.tsx [hunk 14]
```diff
@@ -3,12 +3,14 @@
...
-import { StagePanelLocation, StagePanelSection, StageUsage, UiItemsProvider, Widget } from "@itwin/appui-react";
+import { StagePanelLocation, StagePanelSection, StageUsage } from "@itwin/appui-react";
 import { TreeWidget } from "../TreeWidget";
 import { CategoriesTreeComponent } from "./trees/category-tree/CategoriesTreeComponent";
 import { ModelsTreeComponent } from "./trees/models-tree/ModelsTreeComponent";
-import { TreeDefinition, TreeWidgetComponent } from "./TreeWidgetComponent";
+import { TreeWidgetComponent } from "./TreeWidgetComponent";
+
+import type { UiItemsProvider, Widget } from "@itwin/appui-react";
+import type { TreeDefinition} from "./TreeWidgetComponent";
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/components/TreeWidgetUiItemsProvider.tsx:13]:
```suggestion
import type { TreeDefinition } from "./TreeWidgetComponent";
```

File: packages/itwin/tree-widget/src/components/trees/VisibilityTreeRenderer.tsx [hunk 15]
```diff
@@ -3,23 +3,24 @@
...
+import { useCallback, useEffect } from "react";
+import { TreeImageLoader, TreeNodeRenderer, TreeRenderer } from "@itwin/components-react";
 import { Checkbox } from "@itwin/itwinui-react";
-import { IPresentationTreeDataProvider, useControlledPresentationTreeFiltering } from "@itwin/presentation-components";
-import { VisibilityTreeFilterInfo } from "./Common";
+import { useControlledPresentationTreeFiltering } from "@itwin/presentation-components";
+
+import type {AbstractTreeNodeLoaderWithProvider, TreeNodeRendererProps, TreeRendererProps } from "@itwin/components-react";
+import type { NodeCheckboxRenderProps } from "@itwin/core-react";
+import type { IPresentationTreeDataProvider} from "@itwin/presentation-components";
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/components/trees/VisibilityTreeRenderer.tsx:13]:
```suggestion
import type { IPresentationTreeDataProvider } from "@itwin/presentation-components";
```

File: packages/itwin/tree-widget/src/components/trees/category-tree/CategoriesTree.tsx [hunk 16]
```diff
@@ -4,18 +4,22 @@
...
 import { VisibilityTreeEventHandler } from "../VisibilityTreeEventHandler";
 import { useVisibilityTreeFiltering, useVisibilityTreeRenderer, VisibilityTreeNoFilteredData } from "../VisibilityTreeRenderer";
-import { CategoryInfo, CategoryVisibilityHandler } from "./CategoryVisibilityHandler";
+import { CategoryVisibilityHandler } from "./CategoryVisibilityHandler";
+
+import type { IModelConnection, SpatialViewState, ViewManager, Viewport } from "@itwin/core-frontend";
+import type { Ruleset } from "@itwin/presentation-common";
+import type { IPresentationTreeDataProvider} from "@itwin/presentation-components";
+import type { VisibilityTreeFilterInfo } from "../Common";
+import type { CategoryInfo} from "./CategoryVisibilityHandler";
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/components/trees/category-tree/CategoriesTree.tsx:22]:
```suggestion
import type { IModelConnection, SpatialViewState, ViewManager, Viewport } from "@itwin/core-frontend";
import type { Ruleset } from "@itwin/presentation-common";
import type { IPresentationTreeDataProvider } from "@itwin/presentation-components";
import type { VisibilityTreeFilterInfo } from "../Common";
import type { CategoryInfo } from "./CategoryVisibilityHandler";
```

File: packages/itwin/tree-widget/src/components/trees/category-tree/CategoriesTreeComponent.tsx [hunk 17]
```diff
@@ -4,20 +4,24 @@
...
 import { AutoSizer } from "../../utils/AutoSizer";
-import { CategoryTree, CategoryTreeProps } from "./CategoriesTree";
-import {
-  CategoryInfo, CategoryVisibilityHandler, hideAllCategories, invertAllCategories, showAllCategories, useCategories,
-} from "./CategoryVisibilityHandler";
+import { CategoryTree } from "./CategoriesTree";
+import { CategoryVisibilityHandler, hideAllCategories, invertAllCategories, showAllCategories, useCategories } from "./CategoryVisibilityHandler";
+
+import type { IModelConnection, ScreenViewport } from "@itwin/core-frontend";
+import type { IPresentationTreeDataProvider} from "@itwin/presentation-components";
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/components/trees/category-tree/CategoriesTreeComponent.tsx:21]:
```suggestion
import type { IPresentationTreeDataProvider } from "@itwin/presentation-components";
```

File: packages/itwin/tree-widget/src/components/trees/category-tree/CategoryVisibilityHandler.tsx [hunk 18]
```diff
@@ -3,14 +3,17 @@
...
+import { useAsyncValue } from "@itwin/components-react";
 import { BeEvent } from "@itwin/core-bentley";
-import { IModelApp, IModelConnection, ViewManager, Viewport } from "@itwin/core-frontend";
+import { IModelApp } from "@itwin/core-frontend";
 import { NodeKey } from "@itwin/presentation-common";
 import { isPresentationTreeNodeItem } from "@itwin/presentation-components";
 import { enableCategory, enableSubCategory, loadCategoriesFromViewport } from "../CategoriesVisibilityUtils";
-import { IVisibilityHandler, VisibilityChangeListener, VisibilityStatus } from "../VisibilityTreeEventHandler";
+
+import type { TreeNodeItem} from "@itwin/components-react";
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/components/trees/category-tree/CategoryVisibilityHandler.tsx:14]:
```suggestion
import type { TreeNodeItem } from "@itwin/components-react";
```

**Review by grigasp — CHANGES_REQUESTED** [review]
So many similar formatting issues suggest something's not right with the tooling. I checked out the branch and checked:
- `@typescript-eslint/consistent-type-imports`, when fixing a problem, doesn't add the space, e.g.:
   ```tsx
   import { IFilteredPresentationTreeDataProvider, isPresentationTreeNodeItem } from "@itwin/presentation-components";
   ```
   gets "fixed" to:
   ```tsx
   import type { IFilteredPresentationTreeDataProvider} from "@itwin/presentation-components";
   import { isPresentationTreeNodeItem } from "@itwin/presentation-components";
   ```
- linter doesn't complain about these formatting problems, so doesn't fix them either

Anything we can do about it?

File: packages/itwin/tree-widget/.eslintrc.js (hunk 1 above)
Comment (saskliutas) [reply packages/itwin/tree-widget/.eslintrc.js:9]: I don't think this should cause any issues for apps. This is [React 17 feature](https://legacy.reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html).

Yes this eslint config change is needed to disable eslint rules that require `React` to be imported.

**Review by grigasp — APPROVED** [review]

**Pull request merged**